File: wstore/tst/testwspace.cc
/*
 * ---------------------------------------------------------------------
 *     program testwspace
 * ---------------------------------------------------------------------
 */

#include <iostream>
#include <iomanip>
#include <cmath>
#include <fstream>
#include "QCDNUM/QCDNUM.h"
using namespace std;

void K3(double *table, int (&kk)[5]);
int iP3(double *table, int i, int j, int k);
void K2(double *table, int (&kk)[4]);
int iP2(double *table, int i, int j);
void fillhead(double *obj, double *store);
int  sjekhead(double *hd1, double *hd2);

//----------------------------------------------------------------------
int main() {
    int nwords = 573;
    int ntags  = 7;
    int imi[3], ima[3];
    double w[nwords];
    double qq, *q = &qq;
    
    double head[20], hstore[7][20];
    int iws=0, is1=1, it1=2, it2=3, is2=4, it3=5, it4=6;
    double *pws = &hstore[iws][0];
    double *ps1 = &hstore[is1][0];
    double *pt1 = &hstore[it1][0];
    double *pt2 = &hstore[it2][0];
    double *ps2 = &hstore[is2][0];
    double *pt3 = &hstore[it3][0];
    double *pt4 = &hstore[it4][0];
    
    int ndim   =  2;
    imi[0]     =  1;
    ima[0]     =  5;
    imi[1]     =  1;
    ima[1]     =  5;
    
    cout << endl;
    cout << "=============== init workspace ======================" << endl;
    int iaset1 = WSTORE::iws_WsInit(w,nwords,ntags,"Whatta drag");
    cout << endl;
    cout << "=============== add 1st table  ======================" << endl;
    int iatab1 =  WSTORE::iws_WTable(w,imi,ima,ndim);
    cout << endl;
    cout << "=============== add 2nd table  ======================" << endl;
    int iatab2 =  WSTORE::iws_WTable(w,imi,ima,ndim);
    cout << endl;
    cout << "=============== add 2nd tbset  ======================" << endl;
    int iaset2 = WSTORE::iws_NewSet(w);
    cout << endl;
    cout << "=============== add 3rd table  ======================" << endl;
    int iatab3 =  WSTORE::iws_WTable(w,imi,ima,ndim);
    cout << endl;
    cout << "=============== add 4th table  ======================" << endl;
    int iatab4 =  WSTORE::iws_WTable(w,imi,ima,ndim);

    
//  Store headers
    fillhead(w       ,pws);
    fillhead(w+iaset1,ps1);
    fillhead(w+iatab1,pt1);
    fillhead(w+iatab2,pt2);
    fillhead(w+iaset2,ps2);
    fillhead(w+iatab3,pt3);
    fillhead(w+iatab4,pt4);
        
//  Fill table 1
    int index[] = {imi[0],imi[1]};
    int ia = WSTORE::iws_Tpoint(w,iatab1,index,2)-1;
    for(int j=imi[1]; j<=ima[1]; j++) {
      for(int i=imi[0]; i<=ima[0]; i++) {
        ia += 1;
        w[ia] = double(10*i+j);
      }
    }
//  print headers
    WSTORE::sws_WsHead(w,0);
    WSTORE::sws_WsHead(w,iaset1);
    WSTORE::sws_WsHead(w,iatab1);
    WSTORE::sws_WsHead(w,iatab2);
    WSTORE::sws_WsHead(w,iaset2);
    WSTORE::sws_WsHead(w,iatab3);
    WSTORE::sws_WsHead(w,iatab4);
//  print tree
    WSTORE::sws_WsTree(w);
//  Print table 1
    cout << endl << "Contents of table 1" << endl;
    for(int i=imi[0]; i<=ima[0]; i++) {
        for(int j=imi[1]; j<=ima[1]; j++) {
          ia = iP2(w+iatab1,i,j);
          cout << "  " << int(w[ia]) << " ";
        }
      cout << endl;
    }
    
//  Last table and last set address
    int Ltab  = WSTORE::iws_TFskip(w,iatab3);
    int Lset  = WSTORE::iws_SBskip(w,iaset2);
    cout << endl;
    cout << "ia last  table = " << Ltab + iatab3 << endl;
    cout << "ia first tbset = " << Lset + iaset2 << endl;

    cout << endl;
    cout << "Check WsWipe, TsDump and TsRead" << endl;
    cout << "-------------------------------" << endl;
//  Remove last table
    WSTORE::sws_WsWipe(w,iatab4);
    cout << "Wipe  table 4     " << iatab4 << endl;
//  Dump second table-set
    string fname = "../weights/crap.wgt";
    int ierr;
    WSTORE::sws_TsDump(fname,0,w,iaset2,ierr);
    if(ierr!=0) {
        cout << "Error openening or writing file " << fname << endl;
        return 1;
    }
    else {
        cout << "Dump  tableset 2  " << iaset2 << endl;
    }
//  Wipe second table set
    WSTORE::sws_WsWipe(w,iaset2);
    cout << "Wipe  tableset 2  " << iaset2 << endl;
//  Create new (empty) table set
    int iadum = WSTORE::iws_NewSet(w);
    cout << "New   tableset    " << iadum << endl;
//  Read second table set
    iaset2 = WSTORE::iws_TsRead(fname,0,w,ierr);
    if(ierr!=0) {
        cout << "Error openening or reading file " << fname << endl;
        return 1;
    }
    else {
        cout << "Read  tableset 2  " << iaset2 << endl;
    }
//  Restore last table
    iatab4 = WSTORE::iws_WTable(w,imi,ima,ndim);
    cout << "New   table 4     " << iatab4 << endl;
//  Check headers
    cout << endl;
    fillhead(w,head);
    cout << "workspace header OK = " << sjekhead(head,pws) << endl;
    fillhead(w+iaset1,head);
    cout << "set 1 header     OK = " << sjekhead(head,ps1) << endl;
    fillhead(w+iatab1,head);
    cout << "tab 1 header     OK = " << sjekhead(head,pt1) << endl;
    fillhead(w+iatab2,head);
    cout << "tab 2 header     OK = " << sjekhead(head,pt2) << endl;
    fillhead(w+iaset2,head);
    cout << "set 2 header     OK = " << sjekhead(head,ps2) << endl;
    fillhead(w+iatab3,head);
    cout << "tab 3 header     OK = " << sjekhead(head,pt3) << endl;
    fillhead(w+iatab4,head);
    cout << "tab 4 header     OK = " << sjekhead(head,pt4) << endl;
    
    cout << endl;
    cout << "Check TbCopy" << endl;
    cout << "------------" << endl;
    cout << "Contents of table 4 before copy" << endl;
    for(int i=imi[0]; i<=ima[0]; i++) {
        index[0] = i;
        for(int j=imi[1]; j<=ima[1]; j++) {
            index[1] = j;
            ia = WSTORE::iws_Tpoint(w,iatab4,index,2);
            cout << "  " << int(w[ia]) << " ";
        }
        cout << endl;
    }
    cout << "Copy table 1 to table 4" << endl;
    WSTORE::sws_TbCopy(w,iatab1,w,iatab4,1);
    cout << "Contents of table 4 after copy" << endl;
    for(int i=imi[0]; i<=ima[0]; i++) {
        for(int j=imi[1]; j<=ima[1]; j++) {
            ia = iP2(w+iatab4,i,j);
            cout << "  " << int(w[ia]) << " ";
        }
        cout << endl;
    }
    
    cout << endl;
    cout << "Check WClone" << endl;
    cout << "------------" << endl;
//  Wipe second table set
    WSTORE::sws_WsWipe(w,iaset2);
    cout << "Wipe  tableset 2  " << iaset2 << endl;
//  Create new (empty) table set
    iadum = WSTORE::iws_NewSet(w);
    cout << "New   tableset    " << iadum << endl;
//  Clone table set 1
    iaset2 = WSTORE::iws_WClone(w,iaset1,w);
    cout << "Clone set 1 to 2  " << iaset2 << endl;
//  Wipe table 3
    WSTORE::sws_WsWipe(w,iatab3);
    cout << "Wipe  table 3+4   " << iatab3 << endl;
//  Clone table 2 to 3
    iatab3 = WSTORE::iws_WClone(w,iatab2,w);
    cout << "Clone tab 2 to 3  " << iatab3 << endl;
//  Clone table 3 to 4
    iatab4 = WSTORE::iws_WClone(w,iatab3,w);
    cout << "Clone tab 3 to 4  " << iatab4 << endl;
//  Check headers
    cout << endl;
    fillhead(w,head);
    cout << "workspace header OK = " << sjekhead(head,pws) << endl;
    fillhead(w+iaset2,head);
    cout << "set 2 header     OK = " << sjekhead(head,ps2) << endl;
    fillhead(w+iatab3,head);
    cout << "tab 3 header     OK = " << sjekhead(head,pt3) << endl;
    fillhead(w+iatab4,head);
    cout << "tab 4 header     OK = " << sjekhead(head,pt4) << endl;
    
//  Print tree
    WSTORE::sws_WsTree(w);
    
//  Metadata field
    cout << endl;
    cout << "Fingerprint and metadata table1 with address" << iatab1
         << endl;
    cout << "No-can-do because s/r swsGetMeta is Fortran only" << endl;

//  Build-in pointer function
    cout << endl;
    cout << "Test iws_Tpoint function on table1" << endl;
    index[0] =  imi[0];
    index[1] =  imi[1];
    ia     = WSTORE::iws_Tpoint(w,iatab1,index,2);
    int ib = WSTORE::iws_BeginTbody(w,iatab1);
    cout << "begin table = " << ia << " " << ib << endl;
    index[0] =  ima[0];
    index[1] =  ima[1];
    ia = WSTORE::iws_Tpoint(w,iatab1,index,2);
    ib = WSTORE::iws_EndTbody(w,iatab1);
    cout << "end   table = " << ia << " " << ib << endl;
    
    int icWsp, icSet, icTab;
    WSTORE::sws_WsMark(icWsp,icSet,icTab);
    
    cout << endl;
    cout << "tsize  = " << WSTORE::iws_TbSize(imi,ima,ndim) << endl;
    cout << "hsize  = " << WSTORE::iws_HdSize() << endl;
    cout << "Wspace = " << icWsp << endl;
    cout << "TbSet  = " << icSet << endl;
    cout << "Table  = " << icTab << endl;
    cout << "Yesno  = " << WSTORE::iws_IsaWorkspace(head) << endl;
    cout << "Wsize  = " << WSTORE::iws_SizeOfW(w) << endl;
    cout << "Wused  = " << WSTORE::iws_WordsUsed(w) << endl;
    cout << "Nhead  = " << WSTORE::iws_Nheader(w) << endl;
    cout << "Ntags  = " << WSTORE::iws_Ntags(w) << endl;
    cout << "Hskip  = " << WSTORE::iws_HeadSkip(w) << endl;
    cout << "Drain  = " << WSTORE::iws_IaDrain(w) << endl;
    cout << "INull  = " << WSTORE::iws_IaNull(w) << endl;
    int ityp0 = WSTORE::iws_ObjectType(w,66);
    int ityp1 = WSTORE::iws_ObjectType(w,0);
    int ityp2 = WSTORE::iws_ObjectType(w,iaset1);
    int ityp3 = WSTORE::iws_ObjectType(w,iatab2);
    cout << "Itype  = " << ityp0 << " " << ityp1 << " "
                        << ityp2 << " " << ityp3 << endl;
    cout << "WSsize = " << WSTORE::iws_ObjectSize(w,0) << endl;
    cout << "TSsize = " << WSTORE::iws_ObjectSize(w,iaset1) << endl;
    cout << "TBsize = " << WSTORE::iws_ObjectSize(w,iatab1) << endl;
    cout << "Nsets  = " << WSTORE::iws_Nobjects(w,0) << endl;
    cout << "Ntabs  = " << WSTORE::iws_ObjectNumber(w,iaset1) << " "
                        << WSTORE::iws_Nobjects(w,iaset1) << endl;
    cout << "Ntabs  = " << WSTORE::iws_ObjectNumber(w,iaset2) << " "
                        << WSTORE::iws_Nobjects(w,iaset2) << endl;
    cout << "Wfprnt = " << WSTORE::iws_FingerPrint(w,0) << endl;
    cout << "S1fp   = " << WSTORE::iws_FingerPrint(w,iaset1) << endl;
    cout << "S2fp   = " << WSTORE::iws_FingerPrint(w,iaset2) << endl;
    cout << "T1fp   = " << WSTORE::iws_FingerPrint(w,iatab1) << endl;
    cout << "T2fp   = " << WSTORE::iws_FingerPrint(w,iatab2) << endl;
    cout << "Wtag   = " << WSTORE::iws_IaFirstTag(w,0) << endl;
    cout << "Ndim   = " << WSTORE::iws_TableDim(w,iatab2) << endl;
    cout << "KARR1  = " << WSTORE::iws_IaKARRAY(w,iatab1) << endl;
    cout << "IMIN1  = " << WSTORE::iws_IaIMIN(w,iatab1) << endl;
    cout << "IMAX1  = " << WSTORE::iws_IaIMAX(w,iatab1) << endl;
    cout << "Body1  = " << WSTORE::iws_BeginTbody(w,iatab1) << endl;
    cout << "Bend1  = " << WSTORE::iws_EndTbody(w,iatab1) << endl;
    
    return 0;
}

//----------------------------------------------------------------------

void fillhead(double *obj, double *store) {
    int n = WSTORE::iws_HdSize();
    MBUTIL::smb_vcopy(obj,store,n);
}

int  sjekhead(double *hd1, double *hd2) {
    int n = WSTORE::iws_HdSize();
    return MBUTIL::lmb_vcomp(hd1,hd2,n,-1e-9);
}

void K3(double *table, int (&kk)[5]) {
    int ia    = int(*(table+1));
    double *w = table-ia;
    int iak   = WSTORE::iws_IaKARRAY(w,ia);
    *(kk)     = WSTORE::iws_FingerPrint(w,ia);
    *(kk+1)   = int(*(w+iak));
    *(kk+2)   = int(*(w+iak+1));
    *(kk+3)   = int(*(w+iak+2));
    *(kk+4)   = int(*(w+iak+3));
}

int iP3(double *table, int i, int j, int k) {
    static int kk[5];
    int ia    = int(*(table+1));
    double *w = table-ia;
    int ifp   = WSTORE::iws_FingerPrint(w,ia);
    if( kk[0] != ifp ) { K3( table, kk ); }
    int ip    = kk[1]+kk[2]*i+kk[3]*j+kk[4]*k;
    return ia + ip;
}
                              
void K2(double *table, int (&kk)[4]) {
    int ia    = int(*(table+1));
    double *w = table-ia;
    int iak   = WSTORE::iws_IaKARRAY(w,ia);
    *(kk)     = WSTORE::iws_FingerPrint(w,ia);
    *(kk+1)   = int(*(w+iak));
    *(kk+2)   = int(*(w+iak+1));
    *(kk+3)   = int(*(w+iak+2));
}

int iP2(double *table, int i, int j) {
    static int kk[4];
    int ia    = int(*(table+1));
    double *w = table-ia;
    int ifp   = WSTORE::iws_FingerPrint(w,ia);
    if( kk[0] != ifp ) { K2( table, kk ); }
    int ip    = kk[1]+kk[2]*i+kk[3]*j;
    return ia + ip;
}
                              

File: wstore/tst/testwspace.f

C-----------------------------------------------------------------------
      program testwspace
C-----------------------------------------------------------------------

      implicit double precision (a-h,o-z)
      logical sjekhead

      parameter( nwords = 573 )
      parameter( ntags  = 7 )

      dimension w(nwords)
      dimension imi(2),ima(2)
      dimension kk(27),imin(25),imax(25)
      dimension index(2),row(10)
      character*80 fname
      dimension head(20),hstore(20,7)
      data iSet1/2/, iTab1/3/, iTab2/4/, iSet2/5/, iTab3/6/, iTab4/7/

      write(6,'(/A,I10)') ' WSTORE version = ', iws_Version()

      ndim   =  2
      imi(1) =  1
      ima(1) =  5
      imi(2) =  1
      ima(2) =  5

      write(6,*) ' '
      write(6,*) '=============== init workspace ======================'
      iaSet1 = iws_wsinit(w,nwords,ntags,' ')
      write(6,*) ' '
      write(6,*) '=============== add 1st table  ======================'
      iaTab1 = iws_wtable(w,imi,ima,ndim)
      write(6,*) ' '
      write(6,*) '=============== add 2nd table  ======================'
      iaTab2 = iws_wtable(w,imi,ima,ndim)
      write(6,*) ' '
      write(6,*) '=============== add 2nd tbset  ======================'
      iaSet2 = iws_newset(w)
      write(6,*) ' '
      write(6,*) '=============== add 3rd table  ======================'
      iaTab3 = iws_wtable(w,imi,ima,ndim)
      write(6,*) ' '
      write(6,*) '=============== add 4th table  ======================'
      iaTab4 = iws_wtable(w,imi,ima,ndim)
*      write(6,*) ' '
*      write(6,*) '=============== add 3rd tbset  ======================'
*      iaSet3 = iws_newset(w)
*      call sws_WShead(w,iaSet3)
*      stop

C--   Store headers
      call fillhead(w(1),hstore(1,1))
      call fillhead(w(iaSet1),hstore(1,iSet1))
      call fillhead(w(iaTab1),hstore(1,iTab1))
      call fillhead(w(iaTab2),hstore(1,iTab2))
      call fillhead(w(iaSet2),hstore(1,iSet2))
      call fillhead(w(iaTab3),hstore(1,iTab3))
      call fillhead(w(iaTab4),hstore(1,iTab4))

C--   Fill table 1
      index(1) = imi(1)
      index(2) = imi(2)
      ia = iws_TPoint(w,iaTab1,index,2)-1
      do j = imi(2),ima(2)
        do i = imi(1),ima(1)
          ia    = ia + 1
          w(ia) = dble(10*i+j)
        enddo
      enddo
C--   Print headers
      call sws_WShead(w,1)
      call sws_WShead(w,iaSet1)
      call sws_WShead(w,iaTab1)
      call sws_WShead(w,iaTab2)
      call sws_WShead(w,iaSet2)
      call sws_WShead(w,iaTab3)
      call sws_WShead(w,iaTab4)
C--   Print tree
      call sws_WStree(w)
C--   Print table 1
      write(6,*)," "
      write(6,*) 'Contents of table 1'
      do i = imi(1),ima(1)
        index(1) = i
        do j = imi(2),ima(2)
          index(2) = j
          ia = iws_TPoint(w,iaTab1,index,2)
          row(j) = w(ia)
        enddo
        write(6,*) (int(row(j)),j=imi(2),ima(2))
      enddo

C--   Last table and last set address
      Ltab  = iws_TFskip(w,iaTab3)
      Lset  = iws_SBskip(w,iaSet2)
      write(6,*) ' '
      write(6,*) 'ia last  table = ',Ltab + iaTab3
      write(6,*) 'ia first tbset = ',Lset + iaSet2

      write(6,*) ' '
      write(6,*) 'Check WsWipe, TsDump and TsRead'
      write(6,*) '-------------------------------'
C--   Remove last table
      call sws_WsWipe(w,iaTab4)
      write(6,*)  'Wipe  table 4   ',iaTab4
C--   Dump second table-set
      fname = '../weights/crap.wgt'
      idump = iaSet2
      call sws_TsDump(fname,0,w,idump,ierr)
      if(ierr.ne.0) then
        write(6,*) 'Error openening or writing file ',fname,' '
        stop
      else
        write(6,*) 'Dump  tableset 2',idump
      endif
C--   Wipe second table set
      call sws_WsWipe(w,iaSet2)
      write(6,*)  'Wipe  tableset 2',iaSet2
C--   Create new (empty) table set
      iadum = iws_NewSet(w)
      write(6,*)  'New   tableset  ',iadum
C--   Read second table set
      iaSet2 = iws_TsRead(fname,0,w,ierr)
      if(ierr.ne.0) then
        write(6,*) 'Error openening or reading file ',fname
        stop
      else
        write(6,*)  'Read  tableset 2',iaSet2
      endif
C--   Restore last table
      iaTab4 = iws_WTable(w,imi,ima,ndim)
      write(6,*)  'New   table 4   ',iaTab4
C--   Check headers
      write(6,*) ' '
      call fillhead(w(1),head)
      write(6,*) 'workspace header OK = ',sjekhead(head,hstore(1,1))
      call fillhead(w(iaSet2),head)
      write(6,*) 'set 2 header     OK = ',sjekhead(head,hstore(1,iSet2))
      call fillhead(w(iaTab3),head)
      write(6,*) 'tab 3 header     OK = ',sjekhead(head,hstore(1,iTab3))
      call fillhead(w(iaTab4),head)
      write(6,*) 'tab 4 header     OK = ',sjekhead(head,hstore(1,iTab4))

      write(6,*) ' '
      write(6,*) 'Check TbCopy'
      write(6,*) '------------'
      write(6,*) 'Contents of table 4 before copy'
      do i = imi(1),ima(1)
        index(1) = i
        do j = imi(2),ima(2)
          index(2) = j
          ia = iws_Tpoint(w,iaTab4,index,2)
          row(j) = w(ia)
        enddo
        write(6,*) (int(row(j)),j=imi(2),ima(2))
      enddo
      write(6,*) 'Copy table 1 to table 4'
      call sws_TbCopy(w,iaTab1,w,iaTab4,1)
      write(6,*) 'Contents of table 4 after copy'
      do i = imi(1),ima(1)
        index(1) = i
        do j = imi(2),ima(2)
          index(2) = j
          ia = iws_Tpoint(w,iaTab4,index,2)
          row(j) = w(ia)
        enddo
        write(6,*) (int(row(j)),j=imi(2),ima(2))
      enddo

      write(6,*) ' '
      write(6,*) 'Check WClone'
      write(6,*) '------------'
C--   Wipe second table set
      call sws_WsWipe(w,iaSet2)
      write(6,*)  'Wipe  tableset 2',iaSet2
C--   Create new (empty) table set
      iadum = iws_NewSet(w)
      write(6,*)  'New   tableset  ',iadum
C--   Clone table set 1
      iaSet2 = iws_WClone(w,iaSet1,w)
      write(6,*)  'Clone set 1 to 2',iaSet2
C--   Wipe table 3
      call sws_WsWipe(w,iaTab3)
      write(6,*)  'Wipe  table 3+4 ',iaTab3
C--   Clone table 2 to 3
      iaTab3 = iws_WClone(w,iaTab2,w)
      write(6,*)  'Clone tab 2 to 3',iaTab3
C--   Clone table 3 to 4
      iaTab4 = iws_WClone(w,iaTab3,w)
      write(6,*)  'Clone tab 3 to 4',iaTab4
C--   Check headers
      write(6,*) ' '
      call fillhead(w(1),head)
      write(6,*) 'workspace header OK = ',sjekhead(head,hstore(1,1))
      call fillhead(w(iaSet2),head)
      write(6,*) 'set 2 header     OK = ',sjekhead(head,hstore(1,iSet2))
      call fillhead(w(iaTab3),head)
      write(6,*) 'tab 3 header     OK = ',sjekhead(head,hstore(1,iTab3))
      call fillhead(w(iaTab4),head)
      write(6,*) 'tab 4 header     OK = ',sjekhead(head,hstore(1,iTab4))

C--   Print tree
      call sws_WsTree(w)

C--   Metadata field
      write(6,*) ' '
      write(6,*) 'Fingerprint and metadata table1 with address',iaTab1
      call swsGetMeta(w,iaTab1,nd,kk,imin,imax)
      write(6,'( '' ifp   = '', I15 )') kk(1)
      write(6,'( ''  nd   = '', I15 )') nd
      do i = 1,nd
        write(6,'( '' lim   = '', I15,2i10 )') i,imin(i),imax(i)
      enddo

C--   Build-in pointer function
      write(6,*) ' '
      write(6,*) 'Test iws_Tpoint function on table1'
      index(1) =  imi(1)
      index(2) =  imi(2)
      ia = iws_TPoint(w,iatab1,index,2)
      ib = iws_BeginTBody(w,iatab1)
      write(6,*) 'begin table = ',ia, ib
      index(1) =  ima(1)
      index(2) =  ima(2)
      ia = iws_TPoint(w,iatab1,index,2)
      ib = iws_EndTBody(w,iatab1)
      write(6,*) 'end   table = ',ia, ib

      call sws_WsMark(icWsp,icSet,icTab)

      write(6,*) ' '
      nsize = iws_TbSize(imi,ima,ndim)
      write(6,*) 'tsize  = ', nsize
      nsize = iws_HdSize()
      write(6,*) 'hsize  = ', nsize
      write(6,*) 'Wspace = ', icWsp
      write(6,*) 'TbSet  = ', icSet
      write(6,*) 'Table  = ', icTab
      iyesn = iws_IsaWorkspace(q)
      write(6,*) 'Yesno  = ', iyesn
      nword = iws_SizeOfW(w)
      write(6,*) 'Wsize  = ', nword
      nword = iws_WordsUsed(w)
      write(6,*) 'Wused  = ', nword
      nword = iws_Nheader(w)
      write(6,*) 'Nhead  = ', nword
      nword = iws_Ntags(w)
      write(6,*) 'Ntags  = ', nword
      nword = iws_HeadSkip(w,iaSet2)
      write(6,*) 'Hskip  = ', nword
      nword = iws_IaDrain(w,iatab2)
      write(6,*) 'Drain  = ', nword
      nword = iws_IaNull(w,iaSet1)
      write(6,*) 'Inull  = ', nword
      ityp0 = iws_ObjectType(w,66)
      ityp1 = iws_ObjectType(w,1)
      ityp2 = iws_ObjectType(w,iaSet1)
      ityp3 = iws_ObjectType(w,iaTab2)
      write(6,*) 'Itype  = ', ityp0,ityp1,ityp2,ityp3
      nword = iws_ObjectSize(w,1)
      write(6,*) 'WSsize = ', nword
      nword = iws_ObjectSize(w,iaSet1)
      write(6,*) 'TSsize = ', nword
      nword = iws_ObjectSize(w,iaTab1)
      write(6,*) 'TBsize = ', nword
      nobj  = iws_Nobjects(w,1)
      write(6,*) 'Nsets  = ', nobj
      nobj  = iws_Nobjects(w,iaset1)
      iobj  = iws_ObjectNumber(w,iaSet1)
      write(6,*) 'Ntabs  = ', iobj, nobj
      nobj  = iws_Nobjects(w,iaset2)
      iobj  = iws_ObjectNumber(w,iaSet2)
      write(6,*) 'Ntabs  = ', iobj, nobj
      ifp   = iws_FingerPrint(w,1)
      write(6,*) 'Wfprnt = ', ifp
      ifp   = iws_FingerPrint(w,iaSet1)
      write(6,*) 'S1fp   = ', ifp
      ifp   = iws_FingerPrint(w,iaSet2)
      write(6,*) 'S2fp   = ', ifp
      ifp   = iws_FingerPrint(w,iaTab1)
      write(6,*) 'T1fp   = ', ifp
      ifp   = iws_FingerPrint(w,iaTab2)
      write(6,*) 'T2fp   = ', ifp
      itag  = iws_IaFirstTag(w,1)
      write(6,*) 'Wtag   = ',itag
      mdim  = iws_TableDim(w,iaTab2)
      write(6,*) 'Ndim   = ',mdim
      iakk  = iws_IaKARRAY(w,iaTab1)
      write(6,*) 'KARR1  = ',iakk
      iadr  = iws_IaIMIN(w,iaTab1)
      write(6,*) 'IMIN1  = ',iadr
      iadr  = iws_IaIMAX(w,iaTab1)
      write(6,*) 'IMAX1  = ',iadr
      iadr = iws_BeginTbody(w,iaTab1)
      write(6,*) 'Body1  = ',iadr
      iadr = iws_EndTbody(w,iaTab1)
      write(6,*) 'Bend1  = ',iadr

      end

C     ===========================
      subroutine fillhead(w,head)
C     ===========================

      implicit double precision (a-h,o-z)

      dimension w(*), head(*)

      n = iws_HdSize()
      call smb_Vcopy(w,head,n)

      return
      end

C     ======================================
      logical function sjekhead(head1,head2)
C     ======================================

      implicit double precision (a-h,o-z)
      logical lmb_Vcomp

      dimension head1(*), head2(*)

      n        = iws_HdSize()
      sjekhead = lmb_Vcomp(head1,head2,n,-1.D-9)

      return
      end
